File: Chapter 3/Section 3.1/stamps.cpp
/*
ID: hsfncd31
TASK: stamps
LANG: C++                 
*/
#include <fstream>
#include <vector>
#include <algorithm>
#include <cstring>

static const int MAX_K = 200, MAX_STAMP = 10000;
int min_num_stamp_for_value[MAX_K * MAX_STAMP + 1 + 1];

int main() {
    std::fstream in("stamps.in", std::fstream::in), out("stamps.out", std::fstream::out);

    int k, n;
    std::vector<int> stamps;
    
    in >> k >> n;
    while (stamps.size() < n) {
        int stamp;
        in >> stamp;
        stamps.push_back(stamp);
    }

    std::sort(stamps.begin(), stamps.end());
    memset(min_num_stamp_for_value, -1, (stamps.back() * k + 1 + 1) * sizeof(int));
    min_num_stamp_for_value[0] = 0;
    for (int value = 1; ; ++value) {
        for (auto &stamp: stamps) {
            if (stamp > value)
                break;
            if (min_num_stamp_for_value[value - stamp] != -1
                    && min_num_stamp_for_value[value - stamp] < k
                    && (min_num_stamp_for_value[value] == -1
                        or min_num_stamp_for_value[value - stamp] + 1 < min_num_stamp_for_value[value]))
                min_num_stamp_for_value[value] = min_num_stamp_for_value[value - stamp] + 1;
        }
        if (min_num_stamp_for_value[value] == -1) {
            out << value - 1 << '\n';
            break;
        }
    }

    in.close();
    out.close();
}

File: Chapter 3/Section 3.1/stamps.py
"""
ID: hsfncd31
LANG: PYTHON3
TASK: stamps
Time Limit Exceeded in Case#10
"""
import os
import typing
import itertools


def main():
    base_filename = 'test' if os.name == 'nt' else 'stamps'
    with open('{}.in'.format(base_filename), 'r') as infile,\
            open('{}.out'.format(base_filename), 'w') as outfile:  # type: typing.IO[str]
        def get_line() -> str:
            return infile.readline().rstrip('\n')

        def out_print(*args, **kwargs) -> None:
            kwargs['file'] = outfile
            print(*args, **kwargs)

        k, n = map(int, get_line().split())
        stamps = []
        while len(stamps) < n:
            stamps.extend(map(int, get_line().split()))
        stamps.sort()

        # Looks like humble number in 3.1,
        # but different combination can produce the same result.
        # We want to know the smallest possible number of stamps for each value.
        min_num_stamp_for_value = [-1] * (stamps[-1] * k + 1 + 1)
        min_num_stamp_for_value[0] = 0
        for i in itertools.count(1):
            for stamp in stamps:
                if stamp > i:
                    break
                if min_num_stamp_for_value[i - stamp] != -1\
                        and min_num_stamp_for_value[i - stamp] < k\
                        and (min_num_stamp_for_value[i] == -1
                             or min_num_stamp_for_value[i - stamp] + 1 < min_num_stamp_for_value[i]):
                    min_num_stamp_for_value[i] = min_num_stamp_for_value[i - stamp] + 1
            if min_num_stamp_for_value[i] == -1:
                out_print(i - 1)
                break

if __name__ == '__main__':
    main()
